http4s/server/src/main/scala/endpoints4s/http4s/server/Endpoints.scala

```diff
@@ -361,8 +361,7 @@ trait EndpointsWithCustomErrors extends algebra.EndpointsWithCustomErrors with M
           tupler: Tupler[A, B]
       ): RequestHeaders[tupler.Out] =
         headers =>
-          fa(headers)
-            .flatMap(a => fb(headers).map(b => tupler(a, b)))
+          fa(headers).zip(fb(headers))
```

Comment:
I guess the part of this PR that deals with http4s will be a bit easier to review than the one that deals with akka-http.

I think this is a good example of how structured abstractions make it easier to reason about code… The primary abstraction of akka-http, namely `Directive`, is too powerful IMHO and this makes it hard to reason about code that uses directives.